### ISSUE TITLE
mithril stream .map does not trigger next stream value

### ISSUE THREAD

**Issue by foxdonut** [MithrilJS/mithril.js#2643]
**Mithril version:**

2.0.4

**Browser and OS:**

Chrome on Mac OSX

**Project:**

[Meiosis](https://meiosis.js.org)

## Steps to Reproduce

- Create an `update` stream
- Create a `states` stream by `scan`ning the `update stream
- Call `states.map(state => effect(state)` where `effect` triggers another `update`

## Expected Behavior

- Expect the `effect` function to be called again
- [Here is a flems](https://flems.io/#0=N4IgtglgJlA2CmIBcB2ArAOgMwE4A0IAZhAgM7IDaoAdgIZiJIgYBW5BAxgPbUAu8fZCG7VSvAAQBXAA5Ra-cQF5xhWAE8oGMQCd49ABQBKANwAdaiLHiI1CLwi1YS8cHHauk-knGmQACS4GXzw3D34AYQALWmoAc3gob15tSXgQuV5abwoAXXEAXzNqc0sJMXl4UmdVDS0OGP1zcXF9cv4Q6XkOSMMlAD5xTt5u1sz+Qzwm61t7R0nqZpkM+HMTc3M2yowwWml9aTXiix4reEJCeA4JZU3+lymIQhbNjHdPeCVFZV8AoJBe4BTZqPZ5jeCvMLwKIxeJQAFA5pSWQVUYVO76VwYLGbEJvCLROIJbyERykNLiWBcWhQGyxJIpD75QyHREFcTwWBk6xPVH8DCU6m0+ELVniMm8AAqEAYYX0RjugJFoqRy15H0UAwx4ixWjBIQFNLixNJ5IyWXEFF8UEkHAA1sEfCAAI6SWh23x5Jks0VMoqs-JTAPUQrrURg0jbXZqu5nC5XNXMoogAhkhBXCAnIQAJgALEgAAwgfI5AiwGy28hIKggOgMIQ1KDJkCSbSwISRXi8aSkJAAel7kmo0ltsQw3DAvYbAAF8xgsxgAByT9SaNhN3hqaSMECkDjaCDSXhFks1+jbhtFoA) showing the expected behavior
- Notice the output on the console:
    - `{ route: 'Home', routeChanged: true, data: [] }`
    - `{ route: 'Home', routeChanged: false, data: [], loading: true }`
    - `{ route: 'Home', routeChanged: false, data: [ 'duck', 'quack' ], loading: false }`

## Current Behavior

- The `effect` function is not called again
- [Here is a flems](https://flems.io/#0=N4IgtglgJlA2CmIBcB2ArAOgMwE4A0IAZhAgM7IDaoAdgIZiJIgYBW5BAxgPbUAu8fZCG7VSvAAQBXAA5Ra-cQF5xYDGIBO8egAoAlAG4AOtRFjxEahF4RasJeODj1XSfyTjDIABJcGnvE4u-ADCABa01ADm8FDuvOqS8AFyvLTuFAC64gC+RtTGphJi8vCk9qoaWhUcEdrG4uLaxfwB0vIcobpKAHzibbwdTan8unj15pbWtmPUDTIp8MYGxsbNpRhgtNLa0sv5Jjxm8ISE8BwSyms9DuMQhI1rGM6u8EqKyp4+fiBdwOMNdwew3gTyC8DCEWiUF+-waUlkJSGJWu2kcGHRawCzxC4SiMXchFspCS4lgXFoUAskTiCVe2V0ezhOXE8FgxPM9yR-AwZIpVJhsyZ4mJvAAKhAGEFtHprn9BUL4Qsua9FL1UeJ0WpgQFeZSogSiSSUmlxBRPFBJBwANb+DwgACOklo1s8WXpjKF9LyTOy4191FyK1EwNIGy2yuux1O52VDLyIAIxIQ5wghyEWCQACYsCBshkCLALFbyEgqCA6AwhJBeKF1CQALSVegJkCSdSwIShXi8aSkJAAen7kmo0itkQw3DA-ertYbTbAAAFMxgAAyr-vz1jsEC8ACe0kYIFIHDr0l4ufz5foh5nddgjfiVVzQA) showing the current behavior
- Notice the output on the console:
    - `{ route: 'Home', routeChanged: true, data: [] }`
    - `{ route: 'Home', routeChanged: false, data: [], loading: true }`
- The third state is missing because the `effect` function should have been called again, but it is not.

## Context

This is to set up the [Meiosis](https://meiosis.js.org) pattern with Effects.

*Note*: [This commit](https://github.com/foxdonut/meiosis/commit/9f222a7682b8a057843b08bd4ac0c4834765c037) fixes the same issue in Meiosis-Setup's `simple-stream`.

**Comment by pygy** [comment]
@foxdonut have you investigated how to fix this?

**Comment by foxdonut** [comment]
@pygy yes, see the last line of my description above, there is a link to a commit showing how I fixed it in `simple-stream`. I figure the fix should be similar in mithril stream.

**Comment by pygy** [comment]
Thanks, I'll have to dig into that then :-) (unless you beat me to a PR)

**Comment by ericop** [comment]
@pygy here is a [working flems](https://flems.io/#0=N4IgtglgJlA2CmIBcB2ArAOgMwE4A0IAzvAgMYAu8UyIAFueQA6FID0rArgHaMDWA5hlIB7MK0jlaAJwiwAtIXJT4AQzAABAEwYADLtaLlajACtCIAgDNZ8c0gDaoLmsRIQp8wRFdKPmt8UAAg5GKBVKQIBeQLAMQ1UwAAoASgBuAB0uAPJAiC4IcggVWCjA4ECpYQ5KJED0kAAJUXh6vAqqygBhWhUufipapQ54NrDyFVr7AF1AgF8MrkzswMVw21LY+ONCUl7EzMDAxNXKNsZw0lpkqIA+QPPyS+PxymS8A9z8wuL3rkOQsYtLhpTKZE62DBgFSMRKMEGLLLCLhBeCWSzwCilcG3MofCCWI7gjCVarwKKRaL1JpgFoga7AD6HfGEl7wYkdeDdXr9KD0xmHYKhNbPNY4xLlDCS8FtEldHp9AaBSzFYhtWDCFRQPL8QZSYZzZLwgVzQIkYi5AkiygYdWa7V8v7GlbwcgAFQgNI6iRSOIZjqdgsBVrJkTu4sCkrirLVGq1fVqytgqsCYwmgXs9SgHFIvFadRAAEcOCoc-UZrNDQsAxWqwLZh961x5qDkazCJDocGcaj0RRg5XMhYiCQMYUkXYQJosEgsGg5DokDoQLMpgRYHleHZHCBnDSaBJpLIFEoEkOOFJYDR6EwWOxuHxBCIxAeZPIthptHodAYT8YzEPyAAT0YVwiFIGRGHIIcHkuWxkHsBCdAIfgUlBch0nICs0MbEBV3sOQAEYCEHPCiJAdhDgAeVgKAIyhRhMgo3C8HsJCQAw98O0YQJmM0NA8AAFiQgidAErA8BEsS8NYgh4AASS+IpYAdZj7DIpi8LYjCMIANxUKRAnGKR+ig1dNAEvBNBQCSRPEgjbOktjjAAZQAaTkgAFEiWPU1geM04j0KCgB+QJnKUZiCIANgk-B7IADgshKBMcgg9NgYZDW8tSCA0litKCjDanCqRIpigi4s0ESJKqnRUpAX8kmSbLfP8-LAowoyTLK2KaqsvqUHqxrmoRVTWtUgqMKgeAQK4ahV2i3qCM0KKiOW1ahv0kyRrG3K-ImjryGm2b5ok8rKpwcrNEu+qOEIWhEksYEWr2uoguUchzz+LqXUYvyMMbP7Ai4eAAHc6OhNDOPo0pLG4CgICRR6uDaCB+C4YRlAUgolIdbTyD0gyfpyaISoSVD-TqSniYwc5lB8dtGDuh73xGtCppm+A5q58gybURnmcSYm2aC46uemnwADFkVpwWnpG41+WZZ4jE2VlyUpEAS0KHTaUCAAyfXcnRzH5MU4pAgAQgpMLGowNzPIdANieRxJ0syhW6xbA72Pel0vp6rBNE0CSg7ssOplXYcyDHZEaDQHBF2XKPd1Al8j3fZcgA) based on @foxdonut  example. I tried making a PR, but with other changes in the `next` branch were beyond my wizardry to make play nice together.
